test_catkin_release failing

Getting this failure consistently; I see it in the most recent PRs as well. 

```
======================================================================
FAIL: Release a multi package catkin (melodic) repository.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ubuntu/.local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/ubuntu/esm/ros-common/ros-infrastructure/bloom/test/utils/common.py", line 136, in decorated
    return f(*args, **kwds)
  File "/home/ubuntu/esm/ros-common/ros-infrastructure/bloom/test/system_tests/test_catkin_release.py", line 326, in test_multi_package_repository
    "no debian/melodic/" + pkg + " branch"
AssertionError: no debian/melodic/foo branch

----------------------------------------------------------------------
Ran 1 test in 43.838s

FAILED (failures=1)
```

@cottsay beyond https://github.com/ros-infrastructure/bloom/pull/614 looking like a generally good idea on its own, it also seems to fix this issue. What exactly broke, though?

--- a/.travis.yml
+++ b/.travis.yml
@@ -14,8 +14,6 @@ install:
   - cd /tmp/empy
   - python setup.py install
   - cd -
-  - sudo `which rosdep` init
-  - rosdep update
 # command to run tests
 script:
   - BLOOM_VERBOSE=1 python setup.py nosetests -s --tests test

--- a/test/system_tests/test_catkin_release.py
+++ b/test/system_tests/test_catkin_release.py
@@ -88,7 +88,7 @@ def create_upstream_repository(packages, directory=None, format_versions=None):
         return os.getcwd()
 
 
-def _test_unary_package_repository(release_dir, version, directory=None):
+def _test_unary_package_repository(release_dir, version, directory=None, env=None):
     print("Testing in {0} at version {1}".format(release_dir, version))
     with change_directory(release_dir):
         # First run everything
@@ -98,7 +98,7 @@ def _test_unary_package_repository(release_dir, version, directory=None):
                 cmd = cmd.format(' --quiet')
             else:
                 cmd = cmd.format('')
-            user(cmd, silent=False)
+            user(cmd, silent=False, env=env)
         ###
         ### Import upstream
         ###

--- a/test/system_tests/test_catkin_release.py
+++ b/test/system_tests/test_catkin_release.py
@@ -19,6 +19,7 @@ from .common import create_release_repo
 from ..utils.common import bloom_answer
 from ..utils.common import change_directory
 from ..utils.common import in_temporary_directory
+from ..utils.common import set_up_fake_rosdep
 from ..utils.common import user
 from ..utils.package_version import change_upstream_version
 
@@ -184,6 +185,15 @@ def test_unary_package_repository(directory=None):
     Release a single package catkin (melodic) repository.
     """
     directory = directory if directory is not None else os.getcwd()
+    # Initialize rosdep
+    rosdep_dir = os.path.join(directory, 'foo_rosdep')
+    env = dict(os.environ)
+    fake_distros = {'melodic': {'ubuntu': ['bionic']}}
+    fake_rosdeps = {
+        'catkin': {'ubuntu': []},
+        'roscpp_core': {'ubuntu': []}
+    }
+    env.update(set_up_fake_rosdep(rosdep_dir, fake_distros, fake_rosdeps))
     # Setup
     upstream_dir = create_upstream_repository(['foo'], directory)
     upstream_url = 'file://' + upstream_dir
@@ -198,7 +208,7 @@ def test_unary_package_repository(directory=None):
     versions = ['0.1.0', '0.1.1', '0.2.0']
     import bloom.commands.git.release
     for index in range(len(versions)):
-        _test_unary_package_repository(release_dir, versions[index], directory)
+        _test_unary_package_repository(release_dir, versions[index], directory, env=env)
         bloom.commands.git.release.upstream_repos = {}
         if index != len(versions) - 1:
             change_upstream_version(upstream_dir, versions[index + 1])
@@ -210,6 +220,20 @@ def test_multi_package_repository(directory=None):
     Release a multi package catkin (melodic) repository.
     """
     directory = directory if directory is not None else os.getcwd()
+    # Initialize rosdep
+    rosdep_dir = os.path.join(directory, 'foo_rosdep')
+    env = dict(os.environ)
+    fake_distros = {
+        'melodic': {
+            'debian': ['stretch'],
+            'ubuntu': ['bionic']
+        }
+    }
+    fake_rosdeps = {
+        'catkin': {'debian': [], 'ubuntu': []},
+        'roscpp_core': {'debian': [], 'ubuntu': []}
+    }
+    env.update(set_up_fake_rosdep(rosdep_dir, fake_distros, fake_rosdeps))
     # Setup
     pkgs = ['foo', 'bar_ros', 'baz']
     upstream_dir = create_upstream_repository(pkgs, directory, format_versions=[1, 2, 3])
@@ -230,7 +254,7 @@ def test_multi_package_repository(directory=None):
                 cmd = cmd.format(' --quiet')
             else:
                 cmd = cmd.format('')
-            user(cmd, silent=False)
+            user(cmd, silent=False, env=env)
         ###
         ### Import upstream
         ###
@@ -286,7 +310,7 @@ def test_multi_package_repository(directory=None):
         ### Release generator, again
         ###
         with bloom_answer(bloom_answer.ASSERT_NO_QUESTION):
-            ret = user('git-bloom-generate -y rosrelease melodic -s upstream')
+            ret = user('git-bloom-generate -y rosrelease melodic -s upstream', env=env)
         # patch import should have reported OK
         assert ret == code.OK, "actually returned ({0})".format(ret)
         # Check the environment after the release generator
@@ -371,6 +395,15 @@ def test_upstream_tag_special_tag(directory=None):
     can handle it.
     """
     directory = directory if directory is not None else os.getcwd()
+    # Initialize rosdep
+    rosdep_dir = os.path.join(directory, 'foo_rosdep')
+    env = dict(os.environ)
+    fake_distros = {'melodic': {'ubuntu': ['bionic']}}
+    fake_rosdeps = {
+        'catkin': {'ubuntu': []},
+        'roscpp_core': {'ubuntu': []}
+    }
+    env.update(set_up_fake_rosdep(rosdep_dir, fake_distros, fake_rosdeps))
     # Setup
     upstream_dir = create_upstream_repository(['foo'], directory)
     upstream_url = 'file://' + upstream_dir
@@ -387,4 +420,4 @@ def test_upstream_tag_special_tag(directory=None):
         user('git tag upstream/0.0.0@baz')
 
     import bloom.commands.git.release
-    _test_unary_package_repository(release_dir, '0.1.0', directory)
+    _test_unary_package_repository(release_dir, '0.1.0', directory, env=env)

--- a/test/utils/common.py
+++ b/test/utils/common.py
@@ -124,6 +124,19 @@ class change_directory(object):
             os.chdir(self.original_cwd)
 
 
+class change_environ(object):
+    def __init__(self, env=None):
+        self.original_env = os.environ
+        self.new_env = dict(env) if env is not None else dict(os.environ)
+
+    def __enter__(self):
+        self.original_env = os.environ
+        os.environ = dict(self.new_env)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        os.environ = self.original_env
+
+
 def in_temporary_directory(f):
     @functools.wraps(f)
     def decorated(*args, **kwds):
@@ -169,7 +182,7 @@ class temporary_directory(object):
 
 
 def user_bloom(cmd, args=None, directory=None, auto_assert=True,
-               return_io=True, silent=False):
+               return_io=True, silent=False, env=None):
     """Runs the given bloom cmd ('git-bloom-{cmd}') with the given args"""
     assert type(cmd) == str, \
         "user_bloom cmd takes str only, got " + str(type(cmd))
@@ -191,7 +204,8 @@ def user_bloom(cmd, args=None, directory=None, auto_assert=True,
         with redirected_stdio() as (out, err):
             func = load_entry_point('bloom==' + ver, 'console_scripts', cmd)
             try:
-                ret = func(args) or 0
+                with change_environ(env):
+                    ret = func(args) or 0
             except SystemExit as e:
                 ret = e.code
                 if ret != 0 and auto_assert:
@@ -284,7 +298,7 @@ _special_user_commands = {
 
 
 def user(cmd, directory=None, auto_assert=True, return_io=False,
-         bash_only=False, silent=True):
+         bash_only=False, silent=True, env=None):
     """Used in system tests to emulate a user action"""
     if type(cmd) in [list, tuple]:
         cmd = ' '.join(cmd)
@@ -298,11 +312,12 @@ def user(cmd, directory=None, auto_assert=True, return_io=False,
                     directory=directory,
                     auto_assert=auto_assert,
                     return_io=return_io,
-                    silent=silent
+                    silent=silent,
+                    env=env
                 )
     ret = -1
     try:
-        p = Popen(cmd, shell=True, cwd=directory, stdout=PIPE, stderr=PIPE)
+        p = Popen(cmd, shell=True, cwd=directory, env=env, stdout=PIPE, stderr=PIPE)
         out, err = p.communicate()
         if out is not None and not isinstance(out, str):
             out = out.decode('utf-8')

--- a/test/utils/common.py
+++ b/test/utils/common.py
@@ -20,6 +20,7 @@ except ImportError:
     from io import StringIO
 
 from subprocess import Popen, PIPE, CalledProcessError
+import yaml
 
 
 def assert_raises(exception_classes, callable_obj=None, *args, **kwargs):
@@ -335,3 +336,96 @@ def user(cmd, directory=None, auto_assert=True, return_io=False,
     if return_io:
         return ret, out, err
     return ret
+
+
+def set_up_fake_rosdep(staging_dir, distros={}, rules={}):
+    """
+    Used to create a 'fake' rosdep cache from a locally generated index.
+
+    Example invocation:
+    .. code-block:: python
+
+       env = dict(os.environ)
+       env.update(set_up_fake_rosdep(
+           '/tmp/fake_rosdep',
+           {
+               'melodic': {
+                   'ubuntu': ['bionic']
+               }
+           },
+           {
+               'some_rosdep_key': {
+                   'ubuntu': ['some-package-name']
+               }
+           }))
+
+    :param staging_dir: Scratch directory in which to make infrastructure files.
+    :param distros: Mapping of ROS distributions to populate the index with.
+    :param rules: rosdep rules to populate the rosdep database with.
+
+    :returns: Environment variables which cause Bloom will use the fake cache.
+    """
+    # Construct bare environment
+    rosdistro_dir = os.path.join(staging_dir, 'rosdistro')
+    rosdistro_index_path = os.path.join(rosdistro_dir, 'index.yaml')
+    sources_list_dir = os.path.join(staging_dir, 'sources.list.d')
+    ros_home_dir = os.path.join(staging_dir, 'ros_home')
+    os.makedirs(rosdistro_dir)
+    os.makedirs(sources_list_dir)
+    os.makedirs(ros_home_dir)
+    env = {
+        'BLOOM_SKIP_ROSDEP_UPDATE': '1',
+        'ROSDEP_SOURCE_PATH': os.path.realpath(sources_list_dir),
+        'ROSDISTRO_INDEX_URL': 'file://' + os.path.realpath(rosdistro_index_path),
+        'ROS_HOME': os.path.realpath(ros_home_dir)
+    }
+
+    # Create the specified distributions
+    for distro, platforms in distros.items():
+        distro_dir = os.path.join(rosdistro_dir, distro)
+        distro_yaml_path = os.path.join(distro_dir, 'distribution.yaml')
+        os.makedirs(distro_dir)
+        distro_yaml_data = {
+            'release_platforms': platforms,
+            'repositories': dict({}),
+            'type': 'distribution',
+            'version': 2
+        }
+        with open(distro_yaml_path, 'w') as f:
+            f.write(yaml.dump(distro_yaml_data))
+
+    # Index the specified distributions
+    rosdistro_index_data = {
+        'distributions': {
+            distro: {
+                'distribution': [os.path.join(distro, 'distribution.yaml')],
+                'distribution_cache': 'DOES-NOT-EXIST',
+                'distribution_status': 'active',
+                'distribution_type': 'ros1',
+                'python_version': 2
+            } for distro in distros.keys()
+        },
+        'type': 'index',
+        'version': 4
+    }
+    with open(rosdistro_index_path, 'w') as f:
+        f.write(yaml.dump(rosdistro_index_data))
+
+    # Create the rosdep database
+    rosdep_db_dir = os.path.join(rosdistro_dir, 'rosdep')
+    os.makedirs(rosdep_db_dir)
+    rosdep_db_path = os.path.join(rosdep_db_dir, 'rosdep.yaml')
+    with open(rosdep_db_path, 'w') as f:
+        f.write(yaml.dump(rules))
+
+    # Create the rosdep sources list
+    rosdistro_source_path = os.path.join(sources_list_dir, '50-rosdep.list')
+    with open(rosdistro_source_path, 'w') as f:
+        f.write('yaml file://' + os.path.realpath(rosdep_db_path))
+
+    # Perform the initial rosdep update
+    setup_env = dict(os.environ)
+    setup_env.update(env)
+    user('rosdep update', env=setup_env)
+
+    return env

